implementations/rust/ockam/ockam_api/src/nodes/service/tcp_inlets/session_replacer/mod.rs

```diff
@@ -0,0 +1,449 @@
+
+pub(super) mod selector;
+
+pub(super) struct InletParameters {
+    pub(super) terminal_notifier: TcpInletNotifier,
+    pub(super) outlet_address_selector: OutletMultiAddrSelector,
+    pub(super) prefix_route: Route,
+    pub(super) suffix_route: Route,
+    pub(super) authorized: Option<Identifier>,
+    pub(super) wait_for_outlet_duration: Duration,
+    pub(super) resource: Resource,
+    pub(super) policy_expression: Option<PolicyExpression>,
+    pub(super) secure_channel_identifier: Option<Identifier>,
+    pub(super) disable_tcp_fallback: bool,
+    pub(super) tls_certificate_provider: Option<MultiAddr>,
+    pub(super) enable_nagle: bool,
+    pub(super) skip_handshake: bool,
+}
+
+/// The status of the additional Inlet session.
+pub(super) struct AdditionalInletSessionStatus {
+    pub(super) secure_channel: SecureChannel,
+    pub(super) udp_puncture: UdpPuncture,
+}
+
+/// The status of the Inlet session.
+pub(super) struct InletSessionStatus {
+    pub(super) main_route: Route,
+    pub(super) connection: Connection,
+    pub(super) last_route_key: String,
+    pub(super) additional: Option<AdditionalInletSessionStatus>,
+    pub(super) original_multiaddr: MultiAddr,
+}
+
+pub(super) struct InletSessionReplacer {
+    pub(super) context: Context,
+    pub(super) node_manager: Weak<NodeManager>,
+    pub(super) inlet: Arc<TcpInlet>,
+    pub(super) udp_transport: Option<Arc<UdpTransport>>,
+    pub(super) status: Option<InletSessionStatus>,
+    pub(super) parameters: Arc<InletParameters>,
+}
+
+impl InletSessionReplacer {
+    /// Returns a standalone instance of the replacer, it shared the dependencies but not
+    /// the status
+    pub(super) fn standalone(&self) -> Result<InletSessionReplacer> {
+        Ok(InletSessionReplacer {
+            context: self.context.try_clone()?,
+            node_manager: self.node_manager.clone(),
+            inlet: self.inlet.clone(),
+            udp_transport: self.udp_transport.clone(),
+            parameters: self.parameters.clone(),
+            status: None,
+        })
+    }
+
+    fn udp_puncture_enabled(&self) -> bool {
+        self.udp_transport.is_some()
+    }
+
+    async fn inlet_options(
+        &self,
+        node_manager: &NodeManager,
+        original_multi_addr: &MultiAddr,
+    ) -> Result<TcpInletOptions> {
+        let (incoming_ac, outgoing_ac) = inlet_access_control(
+            &self.context,
+            node_manager,
+            &self.parameters,
+            Some(original_multi_addr),
+        )
+        .await?;
+
+        let options = TcpInletOptions::new()
+            .set_skip_handshake(self.parameters.skip_handshake)
+            .set_enable_nagle(self.parameters.enable_nagle)
+            .with_incoming_access_control(incoming_ac)
+            .with_outgoing_access_control(outgoing_ac);
+
+        let options = if self.udp_puncture_enabled() && self.parameters.disable_tcp_fallback {
+            options.paused()
```

Comment:
Could you add more context about this? Why are udp inlets paused initially if there is no TCP fallback?

Reply:
I've added a comment:

```
// By default, the portal uses the typical route, and switches to UDP puncture once the
// new route gets established.
// But when `tcp_fallback` is disabled, we want *all* traffic to pass through *only*
// the UDP puncture, so we pause the portal until the UDP puncture is complete.
```